File: backend/src/agents/strategist.py
"""Strategist Brain - core trading intelligence agent.

Researches markets, generates strategy ideas, scores backtest results,
and decides portfolio allocation across asset classes.
"""

import json
from typing import Any, Literal, TypedDict

from langchain_core.messages import HumanMessage, SystemMessage
from langchain_xai import ChatXAI

from src.config import get_settings

MARKET_RESEARCH_SYSTEM_PROMPT = """\
You are an elite quantitative market analyst and portfolio strategist with deep \
expertise across crypto, equities, and forex markets.

Your task is to analyze current market conditions and generate actionable trading \
strategy ideas. For each asset class requested, provide:

1. **Market Regime**: Is the market trending (bull/bear), range-bound, or \
transitioning? What is the current volatility regime (low/normal/high/extreme)?

2. **Key Trends & Drivers**: What macro factors, technical patterns, or sentiment \
shifts are driving price action? Identify the dominant narrative.

3. **Opportunities**: Specific trade setups or strategy types that would perform \
well in this regime. Be concrete — name specific indicators, timeframes, and \
asset pairs.

4. **Risk Factors**: What could invalidate the thesis? Key levels to watch, \
upcoming events, correlation risks.

For strategy ideas, think like a systematic trader. Suggest strategies that can be \
coded as Backtrader strategies with clear entry/exit rules:
- Trend-following (moving average crossovers, breakouts, momentum)
- Mean-reversion (RSI extremes, Bollinger Band bounces, pairs trading)
- Volatility strategies (straddles, vol-of-vol, regime switching)
- Statistical arbitrage (correlation breakdown, cointegration)

Output your analysis as JSON with this structure:
{
  "market_analysis": "Your comprehensive market analysis text",
  "strategy_ideas": [
    {
      "name": "Strategy name",
      "asset_class": "crypto|stocks|forex",
      "type": "trend|mean_reversion|volatility|arbitrage",
      "description": "Brief description of the strategy logic",
      "rationale": "Why this strategy suits current conditions"
    }
  ]
}

Generate 3-5 diverse strategy ideas that span different asset classes and strategy \
types to ensure portfolio diversification."""

STRATEGY_GENERATION_SYSTEM_PROMPT = """\
You are an expert algorithmic trading strategy developer specializing in Backtrader \
strategies. Your job is to take a strategy idea and produce a detailed natural \
language specification that can be directly translated into a Backtrader strategy.

For each strategy, provide a DETAILED specification including:

1. **Indicators**: Exact indicators with parameters (e.g., SMA(20), RSI(14), \
Bollinger Bands(20, 2.0))
2. **Entry Rules**: Precise conditions for entering long/short positions. Use \
boolean logic (e.g., "Enter LONG when SMA_fast crosses above SMA_slow AND \
RSI > 50 AND volume > 1.5x average")
3. **Exit Rules**: Conditions for closing positions (take profit, stop loss, \
trailing stop, indicator-based exits)
4. **Position Sizing**: How to size the position (fixed fraction, volatility-based, \
Kelly criterion)
5. **Risk Management**: Stop-loss percentage, maximum position size, maximum \
drawdown before halting
6. **Timeframe**: Candle timeframe (1h, 4h, 1d) and lookback period
7. **Asset**: Specific trading pair or symbol

Be extremely specific. The description must be unambiguous enough that a code \
generator can produce a working Backtrader strategy from it alone.

Output as JSON array:
[
  {
    "name": "Strategy Name",
    "asset_class": "crypto|stocks|forex",
    "description": "Full detailed strategy specification as described above"
  }
]"""

ALLOCATION_SYSTEM_PROMPT = """\
You are a senior portfolio manager making allocation decisions for a systematic \
trading portfolio. You must decide how to distribute capital across a set of \
ranked trading strategies.

Your allocation must respect these hard constraints:
- Maximum {max_position_pct}% of total capital per individual strategy position
- Maximum {max_drawdown_pct}% portfolio-level drawdown tolerance
- Total capital: ${capital:,.0f}

Allocation principles:
1. **Diversification**: Spread across asset classes. Never allocate >40% to a \
single asset class.
2. **Risk Budgeting**: Higher-scoring strategies get more capital, but adjust for \
correlated drawdown risk. Two highly correlated strategies should share a risk budget.
3. **Score Thresholds**: Only allocate to strategies with a composite score above \
0.3. Below that, the risk-reward is unfavorable.
4. **Cash Reserve**: Always keep at least 10% in cash as a risk buffer.
5. **Position Sizing**: Within each strategy's allocation, individual positions \
should not exceed {max_position_pct}% of total portfolio.

Output your decision as JSON:
{{
  "allocations": [
    {{
      "strategy_name": "Name",
      "strategy_id": "id or null",
      "asset_class": "crypto|stocks|forex",
      "capital_pct": 15.0,
      "capital_amount": 15000.0,
      "rationale": "Why this allocation"
    }}
  ],
  "cash_reserve_pct": 10.0,
  "cash_reserve_amount": 10000.0,
  "risk_assessment": "Overall portfolio risk assessment",
  "expected_portfolio_sharpe": 1.5,
  "notes": "Any additional notes on the allocation decision"
}}"""


class StrategyCandidate(TypedDict):
    """A candidate strategy with optional backtest results."""

    description: str
    strategy_id: str | None
    report_id: str | None
    metrics: dict[str, Any] | None
    score: float


class StrategistState(TypedDict):
    """State for the strategist agent pipeline."""

    phase: Literal["research", "generate", "backtest", "score", "allocate", "complete"]
    market_analysis: str | None
    asset_classes: list[str]
    strategy_candidates: list[StrategyCandidate]
    ranked_strategies: list[StrategyCandidate]
    portfolio_allocation: dict[str, Any] | None
    capital: float
    risk_constraints: dict[str, float]
    error: str | None


class StrategistAgent:
    """The trading brain that researches, generates, tests, and selects strategies."""

    def __init__(self) -> None:
        settings = get_settings()
        self.llm = ChatXAI(
            model="grok-2-latest",
            api_key=settings.xai_api_key,
        )
        self.settings = settings

    async def research_markets(self, state: StrategistState) -> StrategistState:
        """Analyze current market conditions across all asset classes.

        Uses LLM to assess market regime, trends, volatility, and generate
        a list of strategy ideas suited to current conditions.
        """
        asset_classes = state["asset_classes"]

        messages = [
            SystemMessage(content=MARKET_RESEARCH_SYSTEM_PROMPT),
            HumanMessage(
                content=(
                    f"Analyze current market conditions for these asset classes: "
                    f"{', '.join(asset_classes)}.\n\n"
                    f"Capital available: ${state['capital']:,.0f}\n"
                    f"Risk constraints: max {state['risk_constraints']['max_position_pct']}% "
                    f"per position, max {state['risk_constraints']['max_drawdown_pct']}% "
                    f"portfolio drawdown.\n\n"
                    f"Generate diverse strategy ideas that work well in current conditions."
                )
            ),
        ]

        response = await self.llm.ainvoke(messages)
        content = str(response.content)

        # Parse the JSON response
        try:
            parsed = json.loads(content)
            state["market_analysis"] = parsed.get("market_analysis", content)
            ideas = parsed.get("strategy_ideas", [])

            # Convert ideas to initial strategy candidates
            candidates: list[StrategyCandidate] = []
            for idea in ideas:
                candidates.append(
                    {
                        "description": (
                            f"{idea.get('name', 'Unknown')}: "
                            f"{idea.get('description', '')} "
                            f"[{idea.get('asset_class', 'unknown')}/"
                            f"{idea.get('type', 'unknown')}] "
                            f"Rationale: {idea.get('rationale', '')}"
                        ),
                        "strategy_id": None,
                        "report_id": None,
                        "metrics": None,
                        "score": 0.0,
                    }
                )
            state["strategy_candidates"] = candidates
        except json.JSONDecodeError:
            # Fallback: use raw response as analysis
            state["market_analysis"] = content
            state["error"] = "Failed to parse market research JSON response"

        return state

    async def generate_strategies(self, state: StrategistState) -> StrategistState:
        """Generate detailed strategy descriptions for each idea.

        Produces Backtrader-compatible strategy specifications that can
        be passed to Lona for code generation.
        """
        candidates = state["strategy_candidates"]
        if not candidates:
            state["error"] = "No strategy candidates to generate from"
            return state

        # Collect the brief descriptions for the LLM
        idea_summaries = "\n".join(
            f"{i + 1}. {c['description']}" for i, c in enumerate(candidates)
        )

        messages = [
            SystemMessage(content=STRATEGY_GENERATION_SYSTEM_PROMPT),
            HumanMessage(
                content=(
                    f"Generate detailed Backtrader strategy specifications for "
                    f"these strategy ideas:\n\n{idea_summaries}\n\n"
                    f"Capital: ${state['capital']:,.0f}\n"
                    f"Max position size: {state['risk_constraints']['max_position_pct']}%\n"
                    f"Max drawdown: {state['risk_constraints']['max_drawdown_pct']}%"
                )
            ),
        ]

        response = await self.llm.ainvoke(messages)
        content = str(response.content)

        try:
            strategies = json.loads(content)
            if isinstance(strategies, list):
                # Update candidates with detailed descriptions
                updated: list[StrategyCandidate] = []
                for i, strat in enumerate(strategies):
                    updated.append(
                        {
                            "description": (
                                f"{strat.get('name', f'Strategy_{i}')}: "
                                f"{strat.get('description', '')}"
                            ),
                            "strategy_id": None,
                            "report_id": None,
                            "metrics": None,
                            "score": 0.0,
                        }
                    )
                state["strategy_candidates"] = updated
        except json.JSONDecodeError:
            state["error"] = "Failed to parse strategy generation JSON response"

        return state

    async def score_strategies(self, state: StrategistState) -> StrategistState:
        """Score and rank strategies based on backtest metrics.

        Scoring formula (composite score 0-1):
        - Sharpe ratio:   40% weight (normalized: sharpe / 3.0, clamped)
        - Max drawdown:   25% weight (inverted: lower is better)
        - Win rate:        20% weight (as decimal 0-1)
        - Total return:   15% weight (normalized: return / 100, clamped)
        """
        candidates = state["strategy_candidates"]
        scored: list[StrategyCandidate] = []

        for candidate in candidates:
            metrics = candidate.get("metrics")
            if not metrics:
                # No backtest results yet — keep with zero score
                scored.append(candidate)
                continue

            sharpe = float(metrics.get("sharpe_ratio", 0.0))
            max_dd = float(metrics.get("max_drawdown", 0.0))
            win_rate = float(metrics.get("win_rate", 0.0))
            total_return = float(metrics.get("total_return", 0.0))

            # Normalize components to 0-1 range
            sharpe_score = min(max(sharpe / 3.0, 0.0), 1.0)
            drawdown_score = min(max(1.0 - (abs(max_dd) / 50.0), 0.0), 1.0)
            win_score = min(max(win_rate / 100.0 if win_rate > 1 else win_rate, 0.0), 1.0)
            return_score = min(max(total_return / 100.0, 0.0), 1.0)

            composite = (
                0.40 * sharpe_score
                + 0.25 * drawdown_score
                + 0.20 * win_score
                + 0.15 * return_score
            )

            candidate["score"] = round(composite, 4)
            scored.append(candidate)

        # Sort by score descending
        ranked = sorted(scored, key=lambda c: c["score"], reverse=True)
        state["ranked_strategies"] = ranked

        return state

    async def decide_allocation(self, state: StrategistState) -> StrategistState:
        """Decide portfolio allocation based on ranked strategies and risk constraints.

        Uses LLM to reason about diversification, risk budget, and position sizing
        to produce a final capital allocation across top strategies.
        """
        ranked = state["ranked_strategies"]
        if not ranked:
            state["error"] = "No ranked strategies to allocate to"
            return state

        # Build strategy summary for LLM
        strategy_summary = "\n".join(
            f"{i + 1}. Score: {s['score']:.4f} | {s['description'][:200]}"
            + (f" | Metrics: {json.dumps(s['metrics'])}" if s.get("metrics") else "")
            for i, s in enumerate(ranked)
        )

        prompt = ALLOCATION_SYSTEM_PROMPT.format(
            max_position_pct=state["risk_constraints"]["max_position_pct"],
            max_drawdown_pct=state["risk_constraints"]["max_drawdown_pct"],
            capital=state["capital"],
        )

        messages = [
            SystemMessage(content=prompt),
            HumanMessage(
                content=(
                    f"Allocate ${state['capital']:,.0f} across these ranked strategies:\n\n"
                    f"{strategy_summary}\n\n"
                    f"Market analysis context:\n{state.get('market_analysis', 'N/A')[:500]}"
                )
            ),
        ]

        response = await self.llm.ainvoke(messages)
        content = str(response.content)

        try:
            allocation = json.loads(content)
            state["portfolio_allocation"] = allocation
        except json.JSONDecodeError:
            # Fallback: store raw text
            state["portfolio_allocation"] = {"raw_response": content}
            state["error"] = "Failed to parse allocation JSON response"

        return state

    async def run(
        self,
        asset_classes: list[str] | None = None,
        capital: float | None = None,
    ) -> dict[str, Any]:
        """Run the full strategist pipeline.

        Phases: research → generate → score → allocate → complete.
        The backtest phase is handled externally via Lona client integration.
        """
        settings = self.settings
        state: StrategistState = {
            "phase": "research",
            "market_analysis": None,
            "asset_classes": asset_classes or ["crypto", "stocks", "forex"],
            "strategy_candidates": [],
            "ranked_strategies": [],
            "portfolio_allocation": None,
            "capital": capital or settings.initial_capital,
            "risk_constraints": {
                "max_position_pct": settings.max_position_pct,
                "max_drawdown_pct": settings.max_drawdown_pct,
            },
            "error": None,
        }

        state = await self.research_markets(state)
        state["phase"] = "generate"

        state = await self.generate_strategies(state)
        # Note: backtest step is called externally via CLI/API
        # since it requires Lona client integration
        state["phase"] = "score"

        state = await self.score_strategies(state)
        state["phase"] = "allocate"

        state = await self.decide_allocation(state)
        state["phase"] = "complete"

        return {
            "market_analysis": state["market_analysis"],
            "strategies": state["ranked_strategies"],
            "allocation": state["portfolio_allocation"],
            "capital": state["capital"],
            "error": state["error"],
        }

File: frontend/src/lib/ai/strategist.ts
import * as ai from 'ai';
import { tool, zodSchema, stepCountIs } from 'ai';
import { xai } from '@ai-sdk/xai';
import { wrapAISDK } from 'langsmith/experimental/vercel';
import { z } from 'zod';

import { getLonaClient } from '@/lib/lona/client';
import type { LonaBacktestRequest, LonaReport } from '@/lib/lona/types';

const { generateText, streamText } = wrapAISDK(ai);
const model = xai('grok-4-1-fast-non-reasoning');

const STRATEGIST_SYSTEM_PROMPT = `You are the Trade Nexus Strategist Brain — an elite quantitative portfolio strategist.

Your mission: research markets, generate trading strategies, backtest them via Lona, score the results, and produce a capital allocation plan.

## Pipeline
1. **Research**: Call research_markets to analyze market conditions for requested asset classes.
2. **Create strategies**: For each promising idea, call create_strategy to generate Backtrader code via Lona AI.
3. **Get market data**: Call list_symbols to check available data. If needed, call download_market_data for missing symbols.
4. **Backtest**: Call run_backtest for each strategy with the relevant data IDs. This returns metrics.
5. **Score**: Call score_strategies with the backtest metrics to rank them.
6. **Allocate**: Call decide_allocation with the ranked strategies and risk constraints.

## Rules
- Always research before creating strategies.
- Create 3-5 diverse strategies spanning different asset classes and types.
- Use provider "xai" when creating strategies for best results.
- When backtesting, use realistic date ranges (at least 6 months of data).
- Only allocate to strategies with a composite score above 0.3.
- Keep at least 10% in cash reserve.
- Never exceed the max position size per strategy.

After completing the full pipeline, summarize: which strategies were created, their backtest scores, and the final allocation plan.`;

const MARKET_RESEARCH_PROMPT = `You are an elite quantitative market analyst. Analyze market conditions and generate strategy ideas.

For each asset class, assess:
1. Market regime (trending/range-bound/transitioning) and volatility regime
2. Key trends, macro drivers, and dominant narratives
3. Specific trade setups for systematic strategies (name indicators, timeframes, pairs)
4. Risk factors that could invalidate the thesis

Suggest strategies codeable as Backtrader strategies with clear rules:
- Trend-following (MA crossovers, breakouts, momentum)
- Mean-reversion (RSI extremes, Bollinger bounces)
- Volatility strategies (regime switching)
- Statistical arbitrage

Output as JSON:
{
  "market_analysis": "comprehensive analysis",
  "strategy_ideas": [
    { "name": "Strategy Name", "asset_class": "crypto|stocks|forex", "type": "trend|mean_reversion|volatility|arbitrage", "description": "strategy logic", "rationale": "why it fits current conditions" }
  ]
}`;

const researchMarketsSchema = z.object({
  assetClasses: z
    .array(z.string())
    .describe('Asset classes to analyze, e.g. ["crypto", "stocks", "forex"]'),
  capital: z.number().describe('Total capital available'),
  maxPositionPct: z
    .number()
    .default(5)
    .describe('Max position size as % of portfolio'),
  maxDrawdownPct: z
    .number()
    .default(15)
    .describe('Max portfolio drawdown tolerance %'),
});

const createStrategySchema = z.object({
  description: z
    .string()
    .min(10)
    .describe('Detailed strategy description with entry/exit rules, indicators, timeframe'),
  name: z.string().optional().describe('Optional strategy name'),
  provider: z
    .string()
    .default('xai')
    .describe('AI provider for code generation (xai recommended)'),
});

const listSymbolsSchema = z.object({
  isGlobal: z
    .boolean()
    .default(false)
    .describe('Only show globally available (pre-loaded) symbols'),
  limit: z.number().default(20).describe('Max symbols to return'),
});

const downloadMarketDataSchema = z.object({
  symbol: z.string().describe('Trading symbol, e.g. BTCUSDT'),
  interval: z
    .string()
    .default('1h')
    .describe('Candle interval: 1m, 5m, 15m, 30m, 1h, 4h, 1d'),
  startDate: z.string().describe('Start date YYYY-MM-DD'),
  endDate: z.string().describe('End date YYYY-MM-DD'),
});

const runBacktestSchema = z.object({
  strategyId: z.string().describe('Lona strategy ID'),
  dataIds: z.array(z.string()).describe('Symbol/data IDs for the backtest'),
  startDate: z.string().describe('Backtest start date YYYY-MM-DD'),
  endDate: z.string().describe('Backtest end date YYYY-MM-DD'),
  initialCash: z.number().default(100000).describe('Starting capital'),
});

const scoreStrategiesSchema = z.object({
  strategies: z.array(
    z.object({
      name: z.string(),
      strategyId: z.string().nullable(),
      reportId: z.string().nullable(),
      metrics: z
        .object({
          sharpe_ratio: z.number().default(0),
          max_drawdown: z.number().default(0),
          win_rate: z.number().default(0),
          total_return: z.number().default(0),
        })
        .nullable(),
    }),
  ),
});

const decideAllocationSchema = z.object({
  rankedStrategies: z.array(
    z.object({
      name: z.string(),
      strategyId: z.string().nullable(),
      score: z.number(),
      assetClass: z.string().optional(),
      metrics: z.record(z.string(), z.unknown()).nullable(),
    }),
  ),
  capital: z.number().describe('Total capital to allocate'),
  maxPositionPct: z.number().default(5).describe('Max % per strategy'),
  maxDrawdownPct: z.number().default(15).describe('Max portfolio drawdown %'),
  marketAnalysis: z.string().optional().describe('Market context summary'),
});

const getMlPredictionSchema = z.object({
  symbol: z.string().describe('Trading symbol'),
  predictionType: z
    .string()
    .default('price')
    .describe('Type: price, anomaly, volatility'),
  timeframe: z.string().default('1d').describe('Prediction timeframe'),
});

const strategistTools = {
  research_markets: tool({
    description:
      'Analyze current market conditions for given asset classes. Returns market regime, key trends, strategy ideas.',
    inputSchema: zodSchema(researchMarketsSchema),
    execute: async (input: z.infer<typeof researchMarketsSchema>) => {
      const { text } = await generateText({
        model,
        system: MARKET_RESEARCH_PROMPT,
        prompt: `Analyze current market conditions for: ${input.assetClasses.join(', ')}.
Capital: $${input.capital.toLocaleString()}.
Risk: max ${input.maxPositionPct}% per position, max ${input.maxDrawdownPct}% drawdown.
Generate 3-5 diverse strategy ideas.`,
      });
      return text;
    },
  }),

  create_strategy: tool({
    description:
      'Create a Backtrader strategy from a natural language description via Lona AI code generation.',
    inputSchema: zodSchema(createStrategySchema),
    execute: async (input: z.infer<typeof createStrategySchema>) => {
      const client = getLonaClient();
      const result = await client.createStrategyFromDescription(
        input.description,
        input.name,
        input.provider,
      );
      return {
        strategyId: result.strategyId,
        name: result.name,
        explanation: result.explanation,
        codePreview: result.code.slice(0, 500) + (result.code.length > 500 ? '...' : ''),
      };
    },
  }),

  list_symbols: tool({
    description:
      'List available market data symbols on Lona. Use isGlobal=true for pre-loaded US equities/forex.',
    inputSchema: zodSchema(listSymbolsSchema),
    execute: async (input: z.infer<typeof listSymbolsSchema>) => {
      const client = getLonaClient();
      const symbols = await client.listSymbols(input.isGlobal, input.limit);
      return symbols.map((s) => ({
        id: s.id,
        name: s.name,
        exchange: s.type_metadata?.exchange ?? 'N/A',
        assetClass: s.type_metadata?.asset_class ?? 'N/A',
        frequencies: s.frequencies,
        isGlobal: s.is_global,
      }));
    },
  }),

  download_market_data: tool({
    description:
      'Download market data from Binance and upload to Lona. Returns the symbol ID for backtesting.',
    inputSchema: zodSchema(downloadMarketDataSchema),
    execute: async (input: z.infer<typeof downloadMarketDataSchema>) => {
      const client = getLonaClient();
      const result = await client.downloadMarketData(
        input.symbol,
        input.interval,
        input.startDate,
        input.endDate,
      );
      return {
        symbolId: result.id,
        name: result.name,
        description: result.description,
      };
    },
  }),

  run_backtest: tool({
    description:
      'Run a backtest for a strategy on Lona. Polls until complete and returns performance metrics.',
    inputSchema: zodSchema(runBacktestSchema),
    execute: async (input: z.infer<typeof runBacktestSchema>) => {
      const client = getLonaClient();
      const request: LonaBacktestRequest = {
        strategy_id: input.strategyId,
        data_ids: input.dataIds,
        start_date: input.startDate,
        end_date: input.endDate,
        simulation_parameters: {
          initial_cash: input.initialCash,
          commission_schema: { commission: 0.001, leverage: 1 },
          buy_on_close: true,
        },
      };

      const { report_id } = await client.runBacktest(request);
      const report: LonaReport = await client.waitForReport(report_id);

      return {
        reportId: report.id,
        status: report.status,
        metrics: report.total_stats,
        error: report.error,
      };
    },
  }),

  score_strategies: tool({
    description: `Score and rank strategies based on backtest metrics.
Composite score (0-1): 40% Sharpe + 25% max drawdown (inverted) + 20% win rate + 15% total return.`,
    inputSchema: zodSchema(scoreStrategiesSchema),
    execute: async (input: z.infer<typeof scoreStrategiesSchema>) => {
      return input.strategies
        .map((s) => {
          if (!s.metrics) return { ...s, score: 0 };

          const sharpeScore = Math.min(Math.max(s.metrics.sharpe_ratio / 3.0, 0), 1);
          const drawdownScore = Math.min(Math.max(1 - Math.abs(s.metrics.max_drawdown) / 50, 0), 1);
          const winScore = Math.min(
            Math.max(s.metrics.win_rate > 1 ? s.metrics.win_rate / 100 : s.metrics.win_rate, 0),
            1,
          );
          const returnScore = Math.min(Math.max(s.metrics.total_return / 100, 0), 1);

          const score =
            0.4 * sharpeScore +
            0.25 * drawdownScore +
            0.2 * winScore +
            0.15 * returnScore;

          return { ...s, score: Math.round(score * 10000) / 10000 };
        })
        .sort((a, b) => b.score - a.score);
    },
  }),

  decide_allocation: tool({
    description:
      'Decide portfolio capital allocation across ranked strategies respecting risk constraints.',
    inputSchema: zodSchema(decideAllocationSchema),
    execute: async (input: z.infer<typeof decideAllocationSchema>) => {
      const strategySummary = input.rankedStrategies
        .map(
          (s, i) =>
            `${i + 1}. Score: ${s.score.toFixed(4)} | ${s.name}${s.metrics ? ` | Metrics: ${JSON.stringify(s.metrics)}` : ''}`,
        )
        .join('\n');

      const { text } = await generateText({
        model,
        system: `You are a senior portfolio manager. Allocate capital across strategies.
Hard constraints: max ${input.maxPositionPct}% per position, max ${input.maxDrawdownPct}% drawdown, total capital $${input.capital.toLocaleString()}.
Only allocate to strategies with score > 0.3. Keep 10%+ in cash.
Output JSON: { allocations: [{ strategy_name, capital_pct, capital_amount, rationale }], cash_reserve_pct, risk_assessment }`,
        prompt: `Ranked strategies:\n${strategySummary}\n\nMarket context:\n${input.marketAnalysis?.slice(0, 500) ?? 'N/A'}`,
      });

      try {
        const jsonMatch = text.match(/\{[\s\S]*\}/);
        if (jsonMatch) return JSON.parse(jsonMatch[0]);
      } catch {
        // fallback
      }
      return { raw_response: text };
    },
  }),

  get_ml_prediction: tool({
    description:
      'Get ML predictions (LSTM, anomaly detection) from the Python backend. Optional enrichment step.',
    inputSchema: zodSchema(getMlPredictionSchema),
    execute: async (input: z.infer<typeof getMlPredictionSchema>) => {
      const mlBackendUrl = process.env.ML_BACKEND_URL ?? 'http://localhost:8000';
      try {
        const url = new URL('/api/predict', mlBackendUrl);
        const response = await fetch(url.toString(), {
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify({
            symbol: input.symbol,
            prediction_type: input.predictionType,
            timeframe: input.timeframe,
          }),
        });

        if (!response.ok) {
          return {
            available: false,
            error: `ML backend returned ${response.status}`,
          };
        }

        return { available: true, ...(await response.json()) };
      } catch {
        return { available: false, error: 'ML backend unavailable' };
      }
    },
  }),
};

export interface StrategistOptions {
  assetClasses?: string[];
  capital?: number;
  maxPositionPct?: number;
  maxDrawdownPct?: number;
}

export interface StrategistResult {
  text: string;
  steps: Array<{
    toolCalls: Array<{ toolName: string; args: unknown }>;
    toolResults: Array<{ toolName: string; result: unknown }>;
    text: string;
  }>;
  usage: { totalTokens: number };
}

function buildPrompt(options: StrategistOptions) {
  const {
    assetClasses = ['crypto', 'stocks', 'forex'],
    capital = 100_000,
    maxPositionPct = 5,
    maxDrawdownPct = 15,
  } = options;

  return `Run the full strategist pipeline.

Asset classes: ${assetClasses.join(', ')}
Capital: $${capital.toLocaleString()}
Risk constraints: max ${maxPositionPct}% per position, max ${maxDrawdownPct}% drawdown.

Start by researching markets, then create strategies, download data if needed, backtest, score, and allocate capital. Produce a complete portfolio plan.`;
}

const sharedConfig = {
  model,
  system: STRATEGIST_SYSTEM_PROMPT,
  tools: strategistTools,
  stopWhen: stepCountIs(20),
} as const;

export async function runStrategist(
  options: StrategistOptions = {},
): Promise<StrategistResult> {
  const result = await generateText({
    ...sharedConfig,
    prompt: buildPrompt(options),
  });

  return {
    text: result.text,
    steps: result.steps.map((step) => ({
      toolCalls: step.toolCalls.map((tc) => ({
        toolName: tc.toolName,
        args: tc.input,
      })),
      toolResults: step.toolResults.map((tr) => ({
        toolName: tr.toolName,
        result: tr.output,
      })),
      text: step.text,
    })),
    usage: { totalTokens: result.totalUsage.totalTokens ?? 0 },
  };
}

export function streamStrategist(options: StrategistOptions = {}) {
  return streamText({
    ...sharedConfig,
    prompt: buildPrompt(options),
  });
}
